### PR TITLE
channel resorting, feature histogram, figure position configurability, and more

### PR DESCRIPTION
**Pull request by agbondy**
Hey Alan! Happy New Year. Here's a PR with a number of changes I've been working on to improve our curation pipeline, some of which I have mentioned to you already. I have tried to incorporate these so they are highly general, in the same "style" as the rest of the repo, and hopefully useful for other users. Let me know what you think!

Significant changes:
1. pull-down menu in main figure provides options for resorting channels by physical location on the probe (either by the first or second dimension of siteLoc) rather than acquisition order. For Neuropixels probes at least, this can be quite useful as the set of channels being addressed can be acquired out of physical order.
2. Addition of a histogram to the figtime plot, as in Phy.
3. User configurability of figure list and positions. 1. "figList" param now respected, which it wasn't before. 2. "figPos" param now added so user can manually set the positions of the figures. Length of figPos must match length of figList if figPos is supplied. If figPos is not supplied, default positions are used for each element of figList. Each element of figPos is a four-element vector containing the normalized figure outer position. (This all needs to be added to the documentation).

Minor changes:
1. If a .prm file exists already in the data directory when importing kilosort output from phy, it is used as a template param file.
2. More useful warning and error messages during param file loading, so that user knows if and how their param file is messed up or will be ignored.

and a variety of smaller bug fixes and improvements.

**N.B.: These changes have been extensively tested for the use case of curating kilosort sorts, but not for curating JRClust sorts. I don't anticipate major problems with that use case since all changes have to do with curation, but I can't rule it out.**

### CODE REVIEW

**Comment by aliddell** [comment]
Happy new year to you, Adrian! I appreciate all the hard work. This should  I'll carve out some time to dig into this PR in the next day or two, but based on your summary, this should take care of what a few people have been asking for for a while now.

**Comment by agbondy** [comment]
Great. Sounds good.

**Comment by vncntprvst** [comment]
Hi @agbondy ,
I tested those changes and I have an error when loading the result file. First I imported KiloSort2 results with `'jrc import-ksort <directory>`. That works fine. Then I ran `jrc 'manual' '<myfile>.prm'`, and I get this error:
> Unrecognized method, property, or field 'spikeRho' for class 'jrclust.sort.TemplateClustering'.

Does that sound familiar? Let me know if you have suggestions on how to proceed. 
Thanks.

**Comment by agbondy** [comment]
I am almost positive it's because "FigRD" is included in the "figList" parameter in your .prm file. This figure relies on data that is only produced by a JRClust sort, and the relevant fields are never created when importing from Kilosort. Try removing that and trying again. I agree we should patch this so it doesn't crash under these conditions -- instead it should throw a warning that the rho-delta figure can't be plotted for kilosort sorts.

**Comment by aliddell** [comment]
I like the FigTime histogram, but is it documented anywhere? Also, I'm getting some strangeness around the figure positions on a JRCLUST sort:

image

Anybody else running into this?

I'm reviewing the other changes. Should have more feedback tomorrow.

File: +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m [hunk 1]
```diff
@@ -0,0 +1,14 @@
+function obj = calculateTemplateSim(obj,flagged,max_lag_s)
+    if nargin<3
+        max_lag_s=5e-4; % 0.5 ms
+        if nargin<2
+            flagged=[];
+        end
+    end
+    max_lag_samples = round(max_lag_s * obj.hCfg.sampleRate);
+    if ~isempty(flagged)
+        obj.templateSim(flagged,flagged) = calculate_simscore(obj.meanWfLocalRaw(:,:,flagged),max_lag_samples,obj.hCfg.siteNeighbors(:,obj.clusterSites(flagged)));
```
Comment (aliddell) [inline +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m]: I don't see `calculate_simscore` defined anywhere.

File: +jrclust/+views/plotFigTime.m [hunk 2]
```diff
@@ -34,24 +57,47 @@
         fgTimes2 = [];
         figTitle = sprintf('Unit %d (black); (press [H] for help)', selected(1));
     end
+   
+    binlimits = [min(bgFeatures) max(bgFeatures)]+eps;     
+    
+    % remove foreground events from background cluster    
+    bg_idx = ~ismember(bgTimes,union(fgTimes,fgTimes2));
```
Comment (aliddell) [inline +jrclust/+views/plotFigTime.m]: Good idea. Might be better in the long run to supply `getFigTimeFeatures` which a parameter of units to skip, but this is fine for now.

File: +jrclust/@Config/validateFigs.m [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+               figPos{f} = [.15 .2 .35 .8];         
+           case 'FigProj'
+               figPos{f} = [.5 .2 .35 .5];              
+           case 'FigSim'
+               figPos{f} = [.5 .7 .35 .3];                            
+           case 'FigRD'   
+               figPos{f} = [.85 0 .15 .25];  
+       end
+    end        
+    if ismember('FigRD',figList)
```
Comment (aliddell) [inline +jrclust/@Config/validateFigs.m]: Here's where the FigRD overlap happens. Your conditional is backward. Here's a correction that (I think) works (I can't verify because `calculate_simscore` is not defined anywhere):

```matlab
function validateFigs(obj)
    %% validates figure list and positions, and fills in with defaults if no positions specified
    if ~isempty(obj.figPos) 
        if numel(obj.figList)~=numel(obj.figPos)
            error('Error parsing params: size of figList does not match size of figPos in param file!');
        end
    else %% use default figure positions
        obj.figPos = defaultFigPos(obj.figList);
    end

    %% add back in required figs if they are missing
    requiredFigs = {'FigWav'};
    if ~all(ismember(requiredFigs,obj.figList))
        warnMsg = sprintf('"%s" will be used despite absence from user-defined figure list.\n',requiredFigs{:}); 
        warndlg(warnMsg,'Missing required figure in "FigList" param');
        for i = 1:length(requiredFigs)
            if ~ismember(requiredFigs{i},obj.figList)
                obj.figList = [obj.figList requiredFigs(i)];
                obj.figPos = [obj.figPos defaultFigPos(requiredFigs{i})];
            end
        end
    end
end

function figPos = defaultFigPos(figList)
    figPos = cell(1, numel(figList));
    hasFigRD = ismember('FigRD', figList);

    for f=1:length(figList)
       switch figList{f}
           case 'FigCorr'
               if hasFigRD
                   figPos{f} = [.85 .25 .15 .25];
               else
                   figPos{f} = [.85 .2 .15 .27];
               end

           case 'FigHist'
               if hasFigRD
                   figPos{f} = [.85 .75 .15 .25];
               else
                   figPos{f} = [.85 .73 .15 .27];
               end

           case 'FigISI'
               if hasFigRD
                   figPos{f} = [.85 .5 .15 .25];
               else
                   figPos{f} = [.85 .47 .15 .26];
               end

           case 'FigMap'
               figPos{f} = [0 .5 .15 .5];

           case 'FigPos'
               figPos{f} = [0 0 .15 .5];

           case 'FigProj'
               figPos{f} = [.5 .2 .35 .5];

           case 'FigRD'
               figPos{f} = [.85 0 .15 .25];

           case 'FigSim'
               figPos{f} = [.5 .7 .35 .3];

           case 'FigTime'
               if hasFigRD
                   figPos{f} = [.15 0 .7 .2];
               else
                   figPos{f} = [.15 0 .85 .2];
               end

           case 'FigWav'
               figPos{f} = [.15 .2 .35 .8];
       end
    end
end
```

File: +jrclust/+detect/@DetectController/detectOneRecording.m [hunk 4]
```diff
@@ -119,48 +119,7 @@
         else
             nSamples = nSamplesLoad;
         end
-
```
Comment (aliddell) [inline +jrclust/+detect/@DetectController/detectOneRecording.m]: I'm having trouble making sense of what's changed here except for skipping a load if it doesn't contain any import times. Everything in the diff is shifted around. Can you summarize?

File: +jrclust/+detect/@DetectController/detectOneRecording.m [hunk 5]
```diff
@@ -171,64 +130,111 @@
             if ~isempty(impSites)
                 iSpikeSites = impSites(inInterval);
             end
-        end
+        end   
+        
+        if ~isempty(impTimes) && ~any(inInterval)
```
Comment (aliddell) [inline +jrclust/+detect/@DetectController/detectOneRecording.m]: You can invert this condition and put everything in the `else` block underneath it, say

```matlab
if isempty impTimes || any(inInterval)
    % load raw samples
    iSamplesRaw = hRec.readRawROI(obj.hCfg.siteMap, 1+loadOffset:loadOffset+nSamples);
    ...
```
```

Comment (agbondy) [reply +jrclust/+detect/@DetectController/detectOneRecording.m]: Done.

File: +jrclust/+detect/@DetectController/detectOneRecording.m (hunk 4 above)
Comment (agbondy) [reply +jrclust/+detect/@DetectController/detectOneRecording.m]: That's all it is. If you don't skip the load but there are no times to import, the code was applying its detection algorithm to find spikes in that load, which is obviously not desirable. For efficiency, I moved the code which determines if there are spikes in that load to the very top of the function so it doesn't even both loading the data for that batch unless it needs to. That's why the code doesn't line up anymore.

File: +jrclust/+curate/@CurateController/updateFigHist.m [hunk 6]
```diff
@@ -18,9 +18,9 @@ function updateFigHist(obj)
         jCluster = selected(2);
     end
 
-    nBinsHist = 50; %TODO: put this in param file (maybe)
+    nBinsHist = 80; %TODO: put this in param file (maybe)
 
-    XData = logspace(0, 4, nBinsHist);
+    XData = logspace(0, 3, nBinsHist);
```
[the remark this answers is not in the document]
Comment (agbondy) [reply +jrclust/+curate/@CurateController/updateFigHist.m]: A personal preference to better see the short time scales which are most indicative for unit quality. I can revert to the default value for the PR.

File: +jrclust/+curate/@CurateController/updateFigHist.m [hunk 7]
```diff
@@ -18,9 +18,9 @@ function updateFigHist(obj)
         jCluster = selected(2);
     end
 
-    nBinsHist = 50; %TODO: put this in param file (maybe)
+    nBinsHist = 80; %TODO: put this in param file (maybe)
```
[the remark this answers is not in the document]
Comment (agbondy) [reply +jrclust/+curate/@CurateController/updateFigHist.m]: Tried it out but I actually prefer 50. Changed it back in the PR.

File: +jrclust/+curate/@CurateController/spawnFigures.m [hunk 8]
```diff
@@ -1,51 +1,40 @@
...
+                figTitle = 'Waveform-based similarity score';                
+            case 'FigHist'
+                figTitle = 'ISI histogram';                
+            case 'FigISI'
+                figTitle = 'Return map';                
+            case 'FigCorr'
+                figTitle = 'Time correlation';                
+            case 'FigRD'
+                if isa(obj.hClust,'jrclust.sort.TemplateClustering')
+                    warning('Skipping spawing of rho-delta plot because density-peak clustering was not used.');
```
[the remark this answers is not in the document]
Comment (agbondy) [reply +jrclust/+curate/@CurateController/spawnFigures.m]: thanks

File: +jrclust/+curate/@CurateController/plotAllFigures.m [hunk 9]
```diff
@@ -61,4 +61,8 @@ function plotAllFigures(obj)
 
     % select first cluster (also plots other figures)
     obj.updateSelect(obj.showSubset(1), 1);
-end
\ No newline at end of file
+
+    % reapply figure positions
```
[the remark this answers is not in the document]
Comment (agbondy) [reply +jrclust/+curate/@CurateController/plotAllFigures.m]: Deleted.

File: +jrclust/+curate/@CurateController/updateFigISI.m [hunk 10]
```diff
@@ -3,8 +3,11 @@ function updateFigISI(obj)
     if isempty(obj.selected) || ~obj.hasFig('FigISI')
         return;
     end
-
-    plotFigISI(obj.hFigs('FigISI'), obj.hClust, obj.hCfg, obj.selected);
+    if numel(double(obj.hClust.spikeTimes(obj.hClust.spikesByCluster{obj.selected(1)})))>1 && (numel(obj.selected)<2 || numel(double(obj.hClust.spikeTimes(obj.hClust.spikesByCluster{obj.selected(2)})))>1)
```
[the remark this answers is not in the document]
Comment (agbondy) [reply +jrclust/+curate/@CurateController/updateFigISI.m]: Yup. That was a result of copying code from elsewhere in the function when something simpler would do fine. Fixed.

File: +jrclust/+views/plotFigTime.m (hunk 2 above)
Comment (agbondy) [reply +jrclust/+views/plotFigTime.m]: Yes that would be more general. Is there a use case for that you can imagine, other than excluding the selected clusters from the background?

File: +jrclust/@Config/validateFigs.m (hunk 3 above)
Comment (agbondy) [reply +jrclust/@Config/validateFigs.m]: Yes looks like you identified the problem. Thanks for the fix!

File: +jrclust/+detect/@DetectController/detectOneRecording.m (hunk 4 above)
Comment (aliddell) [reply +jrclust/+detect/@DetectController/detectOneRecording.m]: > the code was applying its detection algorithm to find spikes in that load, which is obviously not desirable.

Nice catch!

File: +jrclust/+views/plotFigTime.m (hunk 2 above)
Comment (aliddell) [reply +jrclust/+views/plotFigTime.m]: I was thinking just so you could get the full 10000 background spikes, but in most cases a random sample would get you pretty close to that. Otherwise, I can't think of a good use case for that, so we can table that idea for now.

File: +jrclust/+curate/@CurateController/updateFigHist.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (aliddell) [reply +jrclust/+curate/@CurateController/updateFigHist.m]: No, that's a good reason to stick with it. I think you should revert the revert.

**Comment by aliddell** [comment]
@agbondy Is this still WIP, or have you finished making changes?

File: +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m (hunk 1 above)
Comment (agbondy) [reply +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m]: This is the change that was definitely not ready for prime time. Pushed without remembering it would go in the PR. 

That said, I think it's a worthwhile addition once I clean up the code. What `calculate_simscore` does (I'll add this file to the repo) is calculate a similarity score between all clusters based on the mean waveform similarity rather than a template-based similarity score. In my experience, the template-based similarity scores often don't match what you see when you look at the actual spikes. This is for two main reasons I can identify. The first is that Kilosort has the tendency to use more than one template to account for variability in a single cell's spikes. As a result, there can be two templates with low similarity but nearly identical sets of times. From the point of view of spike sorting, I think it makes sense to treat those clusters as having a high similarity, which is better captured by calculating the similarity based on mean waveform rather than the Kilosort templates. Another reason to prefer a similarity metric based on mean waveform is that it is more robust to splitting and merging. After a cluster is split or merged, there's no obvious way to use the templates to calculate similarity anymore. Currently the code just assumes the daughter cells after a split have the same similarity scores as the parent and that a merged cluster has the max similarity with other cells as the two parents. These are obviously not good assumptions. If you recalculate the similarity scores after a split or merge using the newly computed mean waveform, you can avoid these assumptions altogether.

One option is to include code for both a template-based similarity score and a waveform-based one, and include a user-settable parameter that lets you switch between the two.

File: +jrclust/+sort/@TemplateClustering/commit.m [hunk 11]
```diff
@@ -4,44 +4,33 @@
...
-                    templateSim(jCluster, iCluster) = templateSim(iCluster, jCluster);
-                end
-            end
-
-            obj.templateSim = templateSim;
+            
+            
+            obj = obj.calculateTemplateSim();
+            
+%             obj.templatesByCluster = arrayfun(@(iC) unique(obj.spikeTemplates(obj.spikesByCluster{iC})), ...
```
[the remark this answers is not in the document]
Comment (agbondy) [reply +jrclust/+sort/@TemplateClustering/commit.m]: This code has been cleaned up a lot.

File: +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m (hunk 1 above)
Comment (aliddell) [reply +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m]: There's a mean-waveform based similarity metric in vanilla JRCLUST, which can be toggled by pressing 'w' when the similarity plot is active -- or did you have something a bit more sophisticated in mind?

Comment (agbondy) [reply +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m]: I think what I had in mind is likely closely similar. I didn't know about this already existing feature. Let me take a look.

A related question: what are the set of possible use cases of template clustering (that is, cases where the clustering object is of class "jrclust.sort.templateClustering")? I only know of the case of a kilosort import.

File: +jrclust/+curate/@CurateController/updateFigHist.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (agbondy) [reply +jrclust/+curate/@CurateController/updateFigHist.m]: OK done.

File: +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m (hunk 1 above)
Comment (agbondy) [reply +jrclust/+sort/@TemplateClustering/calculateTemplateSim.m]: OK I looked into how JRClust calculates the mean-waveform based similarity metric. It's conceptually very similar to what I was doing, with some minor differences. On the whole, I think what JRClust was already doing was superior so I've undone all my changes related to mean-waveform based similarity metrics. I especially like how the diagonal consists of a "self-similarity" score which can be significantly below 1 when a cluster's waveform is variable. A low self-similarity score seems to be a highly reliable indicator that a cluster that needs splitting! My only suggestion would be to include calculation of a similarity score between clusters with different peak sites. Sometimes two clusters with different peak sites nonetheless need to be merged, and a high similarity score can be a useful indicator. This can be accomplished with a one-character code change, setting the default value of "peaks2" to 1 in jrclust.interfaces.Clustering.computeWaveformSim. I've implemented this change.

**Comment by agbondy** [comment]
OK I think I've addressed all your feedback, Alan. The only thing that isn't done is proper documentation of the new features in the JRClust docs.

**Comment by aliddell** [comment]
Merging. My question isn't that important.

**Comment by agbondy** [comment]
Sounds good! (But wait, which question are you talking about?)
